### ISSUE TITLE
codesign error: Attempted to seek before beginning of stream on file larger than 2GB

### ISSUE THREAD

**Issue by sagecook** [mono/mono#11962]
## Steps to Reproduce

1. make windows executable larger than 2GB
2. try to sign it with codesign on mac
3. 

### Current Behavior

After running for a minute or two I get the error mentioned in the stack trace below.  The file is 2.2G in size so I suspect this is the problem, similar to or related to issue 11000.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:
5.16.0.0

## Stacktrace

```
System.IO.IOException: Attempted to Seek before the beginning of the stream
  at System.IO.FileStream.Seek (System.Int64 offset, System.IO.SeekOrigin origin) [0x0006e] in <98fac219bd4e453693d76fda7bd96ab0>:0
  at Mono.Security.Authenticode.AuthenticodeFormatter.Save (System.String fileName, System.Byte[] asn) [0x0011f] in <21a4abcc52e44877869c07dd17362a52>:0
  at Mono.Security.Authenticode.AuthenticodeFormatter.Sign (System.String fileName) [0x00089] in <21a4abcc52e44877869c07dd17362a52>:0
Couldn't sign file 'xxx.exe'.

```

**Comment by marek-safar** [comment]
Could you please include the whole `codesign` command line

**Comment by sagecook** [comment]
Here is the command we use:
yes <cert key> | $SIGNCODE -spc "${WINOUTPUTPATH}/../signing/authenticode.spc" -v "${WINOUTPUTPATH}/../signing/authenticode.pvk" -a sha1 -$ commercial -n "<product name> Installer" -i "http://www.<company>.com" -t "http://timestamp.verisign.com/scripts/timstamp.dll" -tr 10 "<self extracting windows exe>"

**Comment by sagecook** [comment]
Ugh, I see my paste omitted some values.   Let me try again:
yes our_password | $SIGNCODE -spc "${WINOUTPUTPATH}/../signing/authenticode.spc" -v "${WINOUTPUTPATH}/../signing/authenticode.pvk" -a sha1 -$ commercial -n "${SKUNAME} Installer" -i "http://www.our_company.com" -t "http://timestamp.verisign.com/scripts/timstamp.dll" -tr 10 "${my_selfextracting}" 2>/tmp/signcodeSFX.log

I hope the variable names make sense.

**Comment by marek-safar** [comment]
What is `$SIGNCODE` in your case?

**Comment by sagecook** [comment]
Sorry - I should have provided that.  It is "/Library/Frameworks/Mono.framework/Versions/Current/Commands/signcode.  Here is more info:

3:simcur@MacMini:~ $/Library/Frameworks/Mono.framework/Versions/Current/Commands/signcode
Mono SignCode - version 5.16.0.0
Sign assemblies and PE files using Authenticode(tm).
Copyright 2002, 2003 Motus Technologies. Copyright 2004-2008 Novell. BSD licensed.

**Comment by sagecook** [comment]
I was reviewing your releases and see that I am downrev a bit.  I wonder if I should simply upgrade to the latest and see if that works.  If, for example, the newer code has been compiled with Uint variables that might address this seek message.

**Comment by sagecook** [comment]
BTW, I tried signcode version 5.18 but got the same error.

**Comment by rjvdboon** [comment]
In current master in `AuthenticodeFormatter.Save` the filesize is [cast to an int32](https://github.com/mono/mono/blob/master/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeFormatter.cs#L268), and subsequently used in the [Seek](https://github.com/mono/mono/blob/master/mcs/class/Mono.Security/Mono.Security.Authenticode/AuthenticodeFormatter.cs#L294). Looking further it seems that the offset to where the certificat/signature is stored as dWORD in the PE header, thus uint32.
Have you tried signing using .Net framework's signcode on a Windows machine?
